### PR TITLE
Start of Bluetooth Handsfree rework

### PR DESCRIPTION
**Pull request by lkernan**
This reworks the BonovoBluetooth handsfree phone app in the following ways:
- Replaces the old Boolean method of switching views with one based on enum values.  It should be easier to maintain in future.
- Adds a full screen phone ringing screen without the dial buttons. 
  It's not very pretty yet, i've gone for function over looks so far.  Bigger text and buttons make it easier to answer a call when driving.
- Adds a button on the in call screen to show or hide the dialing buttons.  Allows for phone systems where you have to press numbers in a call.  
  The button to go back from the dialer to the call is the back facing arrow in the top left corner.

Future work will include adding in call buttons for holding and resuming calls, handling call waiting and conference calls, etc.
There shouldn't be any loss of functionality over the original system at this stage.

### CODE REVIEW

File: BonovoBluetooth/src/com/bonovo/bluetooth/BonovoBluetoothHandfree.java [hunk 1]
```diff
@@ -359,49 +387,59 @@ protected void onResume() {
 			}
 		}
 		mDialpadPressCount = 0;
+
```
Comment (jamesgol) [inline BonovoBluetooth/src/com/bonovo/bluetooth/BonovoBluetoothHandfree.java:496]: Can we try not to make so many whitespace changes like these?  In the future if there is another source leak it just makes it a bit harder to do a comparison of changes.

**Comment by jamesgol** [comment]
I haven't even tested any bluetooth calling on my setup yet so I can't test if it works, but on a quick glance everything looks good.  I think it's great to have someone else interested in working on this.

File: BonovoBluetooth/src/com/bonovo/bluetooth/BonovoBluetoothHandfree.java (hunk 1 above)
Comment (lkernan) [inline BonovoBluetooth/src/com/bonovo/bluetooth/BonovoBluetoothHandfree.java:496]: Sorry, i'll try to remember not to do that again.
Had a problem tonight with the Chinese text in one of the CPP files getting garbled.  Sorry for the noise but I'm pretty sure everything is back in order now.

Comment (lkernan) [inline BonovoBluetooth/src/com/bonovo/bluetooth/BonovoBluetoothHandfree.java:496]: I need to test my latest changes in the car yet, but everything compiles and runs in the house.  

All the JNI code should be in place for call waiting, getting Bluetooth media information (track name, artist, current position, etc) and sending custom commands to the phone via SPP.

In the next day or two i'll try to get the call waiting service code connected to the UI and it should be looking pretty good.

Comment (lkernan) [inline BonovoBluetooth/src/com/bonovo/bluetooth/BonovoBluetoothHandfree.java:496]: Ok, so I got a chance to test in the car as I drove to work this morning, and it crashes as soon as the phone connects.. 
I'll move the computer out to the garage tonight and sort it out. I'm jealous of you guys with more than one headunit, testing would be a lot easier inside :-)

Comment (jamesgol) [inline BonovoBluetooth/src/com/bonovo/bluetooth/BonovoBluetoothHandfree.java:496]: I actually did a test of the code earlier today and made a couple outgoing calls and it didn't crash for me.

**Comment by jamesgol** [comment]
This looks like it's shaping up nicely.  Can you post an APK the the forum for people to test as you keep making changes?  I'll try to test some of the stuff myself also.  When you are changes are completed we can merge this in and put out another release.

**Comment by lkernan** [comment]
Yes, we're getting close to a point where I can put an APK out.  I'd just like to get the call waiting stuff finished (hopefully tonight).  
I've just about finished writing the service side intent broadcast and receivers, just need to hook them up to the UI side and test.

**Comment by jamesgol** [comment]
Great!  We are going to split up this repository into the individual elements and keep the original intact to track past and future source leaks.  I think after your changes are finished we can commit them and then do the split up.  I think it will be a better long term strategy to have each of them split into their own repository.

**Comment by lkernan** [comment]
One question.  I tried building just an APK file by doing "make BonovoBluetooth". 
I got an APK file in the out folder which claimed to have installed, but doesn't seem to have, it was still the previous version.  Up to now i've been building update ZIP files and that works.

Am i doing something wrong or is it just not possible to update the built in apps in that method?

I'm just doing a few cleanups on the code now, call waiting works and you can swap between the active and inactive calls fine. The UI isn't very pretty but that is quite easy to fix after the basics are tested a bit.

**Comment by Zaphod-Beeblebrox** [comment]
Your make command should have worked fine, and it sounds like the new apk was placed in /out/system/app/ ??  If so, then it built right.

How did you push this to the android box?  You may not have luck 'installing it'.  You probably need to remove the original apk, and put the new one in /system/app/

Will require remounting /system.  I do this from the command line thus:

adb remount
adb push {newapk} /system/app/{newapk}
adb chmod 644 /system/app/{newapk}
adb reboot

**Comment by lkernan** [comment]
Ok, that's what i did wrong, i put it on a USB stick and tried installing it via ES explorer.
Thanks for the tip.

**Comment by Zaphod-Beeblebrox** [comment]
I just took a very quick look at a few commits, so please excuse me if I overlooked something.  But I see where you switched to AudioFocus (which I think will be a better solution), but I don't see where you are releasing the focus?  Need to do that on call end I think.

**Comment by Zaphod-Beeblebrox** [comment]
Nevermind.  Just saw it .. there's a lot of code changes in here :)  I've not gone through much of it.

**Comment by lkernan** [comment]
:-)  abandonAudioFocus();
Get called when phonestate switches to idle in a couple of places.

**Comment by lkernan** [comment]
Ok, i've just pushed my latest changes across.  I might hold off sending out an APK on the forums until tomorrow as it's getting late here and there are still a couple of minor issues i'd like to check and see if i can address.

Please feel free to give it a go, all the basic phone functions seem pretty stable now.  Call waiting has had only basic testing but works for me...

It seems a little touchy about connecting to the phone on first run after upgrading. Toggling the phone option in the settings screen gets it working again.

For some reason it's not closing when the call is over which i'd like to check.  Google Play music isn't continuing after the call so i'll check that too.  

Here is a link to the APK file on dropbox
https://www.dropbox.com/s/qavfo9f52efkae1/BonovoBluetooth.apk?dl=0 

If it crashes on a call the call will probably continue on without the UI.  In that case your phone will be able to control the call and end it ok.

**Comment by lkernan** [comment]
Ok I think I found why music wasn't resuming.  It was passing null to the abandonAudioFocus call instead of the listener.  
I'll try a fresh build on my drive to work shortly and if it works ok i'll put that build out as a test version.

**Comment by lkernan** [comment]
How do you keep these separate?  The music sleeping change is good to go, but I expected to keep it separate from the bluetooth changes.

**Comment by Zaphod-Beeblebrox** [comment]
Branches are the way to do it, but you probably shouldn't have merged your musicsleeping branch back to master.  Should issue a pull request directly from that branch to nu3001/master.

The other thing I would suggest is setting up another branch for your UI re-write code and pushing it to your github, but not a pull request to NU3001.

When you are happy with all of the work, you should learn how to rebase your commit down to one single commit, push that to your master branch, and then pull request to nu3001.

What ends up happening there is you push one single commit to the main repo, rather than all these little ones as you worked it out.
https://www.atlassian.com/git/tutorials/rewriting-history

**Comment by lkernan** [comment]
Closed as i'm trying to clean up my github tree.  I've got a working branch and have reset master to nu3001 master.  From now on i'll work on the branch and push to my master.  Will create a new pull request from there later.

**Pull request closed without merging**